### ISSUE TITLE
Building UI - seems like unclean setup

### ISSUE THREAD

**Issue by kittkatter** [safing/portmaster-ui#162]
**What happened**:
UI Build on Linux (Manjaro - Arch)
bash pack
```
build list:
[app-electron] linux_amd64 v0.2.2
[app-electron] windows_amd64 v0.2.2
[notifier] linux_amd64 v0.2.3
[notifier] windows_amd64 v0.2.3
[modules] base v0.2.11 (ignoring)
[modules] console v0.1.11 (ignoring)
[modules] monitor v0.2.4 (ignoring)
[modules] portmaster v0.1.13 
[assets] v0.2.2

press [Enter] to start building



[....] ./app-electron/pack build
grep: /home/user/portmaster-ui/app-electron/dist/latest-linux.yml: No such file or directory
[app-electron] linux_amd64 v0.2.2: Cannot find built version 0.2.2, please build manually.
[FAIL] ./app-electron/pack build
[CRIT] ABORTING...
```

 electron-builder
``` 
• electron-builder  version=22.11.7 os=5.10.56-1-MANJARO
 • description is missed in the package.json  appPackageFile=/home/user/Downloads/portmaster-ui-app-electron-v0.2.2/package.json
  • author is missed in the package.json  appPackageFile=/home/user/Downloads/portmaster-ui-app-electron-v0.2.2/package.json
  ⨯ Please specify 'name' in the package.json (/home/user/Downloads/portmaster-ui-app-electron-v0.2.2/package.json)
Please specify 'version' in the package.json (/home/user/Downloads/portmaster-ui-app-electron-v0.2.2/package.json)
```

I created the file with version info now this part of the build works. angular making problems now please check commend below

**What did you expect to happen?**:
Easy to build UI script, within docker I'm getting the same error and sadly I don't know a lot of Electron and couldn't find a solution online to fix this issue. 

**please build manually and the used paths** 
are unclear to me, what I have to do in order to fix this problem.

Please be so kind to help the community grow with a more inclusive building possibility for the UI. I like to try adding some features. Everything beside the UI  I'm able to build fine.

this is not helping me, because how shall i know the content of the "./dist/latest-linux.yml"
https://github.com/safing/portmaster-ui/blob/develop/app-electron/pack

```

  # check if version is built (needs manual building)
  if [[ $(grep "$version" "${baseDir}/dist/latest-linux.yml") == "" ]]; then
    echo -e "\n${COL_BOLD}[app-electron] $platform v$version: ${COL_RED}Cannot find built version $version, please build manually.${COL_OFF}"
    exit 1
  fi
```

**Comment by kittkatter** [comment]
```
 OK ] ./notifier/pack build

[....] ./modules/pack build
base is being ignored, skipping...
console is being ignored, skipping...
monitor is being ignored, skipping...

packaging portmaster


> portmaster@0.1.13 build
> NODE_ENV=production ng build --prod --base-href /ui/modules/portmaster/ --deploy-url /ui/modules/portmaster/

Option "--prod" is deprecated: Use "--configuration production" instead.
Option "extractCss" is deprecated: Deprecated since version 11.0. No longer required to disable CSS extraction for HMR.
You seem to not be depending on "@angular/core" and/or "rxjs". This is an error.

[modules] portmaster v0.1.13: BUILD FAILED.
[FAIL] ./modules/pack build
[CRIT] ABORTING..
```

**Comment by kittkatter** [comment]
```
build list:

[modules] base v0.2.11 (ignoring)
[modules] console v0.1.11 (ignoring)
[modules] monitor v0.2.4 (ignoring)
[modules] portmaster v0.1.13 

press [Enter] to start building

base is being ignored, skipping...
console is being ignored, skipping...
monitor is being ignored, skipping...

packaging portmaster


> portmaster@0.1.13 build /home/user/test/portmaster-ui/modules/portmaster
> NODE_ENV=production ng build --prod --base-href /ui/modules/portmaster/ --deploy-url /ui/modules/portmaster/

Option "--prod" is deprecated: Use "--configuration production" instead.
Option "extractCss" is deprecated: Deprecated since version 11.0. No longer required to disable CSS extraction for HMR.
⠋ Generating browser application bundles (phase: setup)...An unhandled exception occurred: Script file node_modules/marked/lib/marked.js does not exist.
See "/tmp/ng-vLa64I/angular-errors.log" for further details.
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! portmaster@0.1.13 build: `NODE_ENV=production ng build --prod --base-href /ui/modules/portmaster/ --deploy-url /ui/modules/portmaster/`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the portmaster@0.1.13 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/user/.npm/_logs/2021-08-19T07_46_54_939Z-debug.log

[modules] portmaster v0.1.13: BUILD FAILED.

```

**Comment by kittkatter** [comment]
Please please make a clean building setup where it builds the source given in the folder, i can see  portmaster-app_v0-2-2.zip seems to be downloaded again. how shall someone be able to reverse engineer all used dependencies and  build this, fix all building scripts to local sources etc. Please fix these building, otherwise I and many others are unable to help you to improve the project because its way to much work before we can start working dealing with this building setup. If I'm not smart enough and got something wrong please let me know, sadly this will be a quick end of support. But i would really like to comeback trying it again at a later time.

**Comment by kittkatter** [comment]
if i do npm install in the subdirectory in modules i get a lost of warnings.

```
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
npm WARN deprecated babel-eslint@10.1.0: babel-eslint is now @babel/eslint-parser. This package will no longer receive updates.
npm WARN deprecated @hapi/topo@3.1.6: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated @hapi/bourne@1.3.2: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated @hapi/address@2.1.4: Moved to 'npm install @sideway/address'
npm WARN deprecated @hapi/hoek@8.5.1: This version has been deprecated and is no longer supported or maintained
npm WARN deprecated @hapi/joi@15.1.1: Switch to 'npm install joi'
npm WARN deprecated tar@2.2.2: This version of tar is no longer supported, and will not receive security updates. Please upgrade asap.
npm WARN deprecated highlight.js@9.18.1: Version no longer supported. Upgrade to
```

**Comment by kittkatter** [comment]
Option "--prod" is deprecated: Use "--configuration production" instead.
Option "extractCss" is deprecated: Deprecated since version 11.0. No longer required to disable CSS extraction for HMR.
⠋ Generating browser application bundles (phase: setup)...Compiling @angular/core : es2015 as esm2015
Compiling @angular/animations : es2015 as esm2015
Compiling @angular/cdk/keycodes : es2015 as esm2015



[assets] v0.2.2: successfully built.
[ OK ] ./assets/pack build

finished building.




still sadly i can say this is a very very weird building progress with alot messed up, i cant confirm if the build is really working because i got alot of npm errors before when installing the deps, we will see. rly sad since the portmaster-core is a very nice tool.

**Comment by dhaavi** [comment]
Hey @kittkatter, thanks for trying to build and wanting to contribute to the Portmaster!

The Portmaster UI conists of multiple components. It's true that it's not that easy to build, that is why the pack scripts exist. And you are right, there is room for improvement.

Building with the pack scripts should actually work fine - at least it does for us.

Electron is a bit weird, so it is built in a two step process. For that you need build electron app before running the pack scipts: `./pack-container.sh npm run pack` in `app-electron`.
Sorry for not mentioning this.

The portmaster module should have built fine though. We are currently releasing the next version and we had some problems with a conflicting PR, which we will resolve soon. This means that this part may not build currently. Normally, the master branch should always build.

> Please please make a clean building setup where it builds the source given in the folder, i can see portmaster-app_v0-2-2.zip seems to be downloaded again.

If you want the Portmaster to use assets you compiled yourself, you'll need to either manually drop them into the updates directory. If in [Development Mode](https://docs.safing.io/portmaster/settings#core/devMode), you can also use the suffix `_v0-0-0`). 

> why are they ignored?

These are modules from an older version of the UI. Some of them still work and used sometimes for debugging, but are currently not released or advertised.

> how shall i know what is needed for a simple rebuild of the ui...

We will improve the documentation for that.

> i just want to check if little code changes work like intended...

In that case, you can just run `./docker.sh` in `modules/portmaster`.
Then, enable [Development Mode](https://docs.safing.io/portmaster/settings#core/devMode) and access the UI dev server in your browser at `http://127.0.0.1:8081`.

**Comment by kittkatter** [comment]
i got everything build now, but the npm-all in ./modules is not working.

i had to manually go into every dir (otherwise final build would not work)
and use the npm install
this was really confusing. beside this the dependencies are not obvious and the reliance on other git repos is very disturbing because if anything goes offline / the repo gets moved it will mess with all the scripts you build. 

it would be really nice if you could update the readme relating towards `./pack-container.sh npm run pack`
if you like i could edit the readme by myself and you can check which content you think fits the repo. 

I don't want to mess it up with to much info but the amount of barriers i had to pass in order to get this build was quite a challenge, especially if you don't work with npm and electron often.

**Comment by davegson** [comment]
> Please be so kind to help the community grow with a more inclusive building possibility for the UI. I like to try adding some features. Everything beside the UI I'm able to build fine.
>  I have to admit this repo does not invite people to contribute.

Hey @kittkatter, thank you so much for fighting through the messy jungle! The PM UI jungle seems to be much denser than what we had hoped for. Definitely want to improve, and if you could propose changes to the readme which include your learnings that would be amazing!

> I don't want to mess it up with too much info

I feel the more info the better, but I would much rather cut things down or move them to a sub-paragraph than to miss something which might turn out to be super valuable for the next person.

Thanks again! Have a good one, David

PS: the release @dhaavi talked about got postponed to next week since we ran into an unexpected blocker

**Comment by kittkatter** [comment]
I  was on vacation. 

Sadly it's quite much to do with all the weird dependencies.
I suggest to add 2 basic scripts for easier building.
I was motivated in the beginning than my ubuntu vm crashed and i did not wont to research the needed packages so the scripts will be arch related only.

One important question i have. Once i package every part of the portmaster it keeps auto updating therefore i cant check if the version works as intended. **Is there a easy was to disable the portmaster-core from auto updating ?** Because i keeps pulling official builds later on other because you have bind this with the json request to your own server and i was not quickly able to disable this. I hope there is some global setting for this, please help me out if your interested in my input.
If i have the time, i will add build deps and instructions for the other parts of the project.
I  interested to help adding some dns related features. Because some Costumers will have a Router which handles this filtering process and would be bypassed, also a lock down mode for inexperienced user is interesting since social engineering is being one huge threat for sys admins. 

**dependencies.sh**

```
#!/bin/sh
# has to be posix, so bash, dash, zsh etc no fish!
sudo pacman -Syy
sudo pacman -S --needed jq npm docker  yay  base-devel libnetfilter_queue --noconfirm
sudo systemctl enable docker
# but  sudo usermod -aG  docker user needs to be executed or run docker as sudo
# since $(who) does not give a nice output
#sudo usermod -aG  docker $(ls /home)
yay -S --noconfirm media-control-indicator-git
#its the appindicator3

sudo npm install -g @angular/cli@latest

mkdir ./app-electron/dist
echo '$version: 0.2.2' > ./app-electron/dist/latest-linux.yml

cd ./modules/base
npm install
cd ..
cd ./console
npm install
cd ..
cd ./monitor
npm install
cd ..
cd ./portmaster
npm install
cd ..
cd ..
```


**build.sh**

```
#!/bin/sh
# source dir can not contains () or other special chars
cd app-electron
sudo sh ./pack-container.sh npm run pack
cd ..
sudo sh ./pack

echo 'Output will be in:

./dist/all/ui/modules/assets_v0-2-2.zip
./dist/all/ui/modules/portmaster_v0-1-14.zip
./dist/linux_amd64/app/portmaster-app_v0-2-2.zip
./dist/linux_amd64/notifier/portmaster-notifier_v0-2-3'
```
